### ISSUE TITLE
TypeError: step() missing 1 required positional argument: 'step'

### ISSUE THREAD

**Issue by Masanori-Yoshihira** [ethnhe/PVN3D#69]
I encountered the error shown in the title when I ran the following code

$ python3 -m train.train_ycb_pvn3d
/Programs/PVN3D/pvn3d/common.py:134: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  self.lm_r_lst = yaml.load(lm_r_file)
train_dataset_size:  96189
Loading valtestset.
Finish loading valtestset.
val_dataset_size:  2949
loading pretrained mdl.
Traceback (most recent call last):
  File "train/train_ycb_pvn3d.py", line 477, in <module>
    mode = 'triangular'
  File "/Programs/PVN3D/pvn3d/lib/utils/warmup_scheduler/scheduler.py", line 240, in __init__
    super().__init__(optimizer)
  File "/usr/local/lib/python3.6/dist-packages/torch/optim/lr_scheduler.py", line 78, in __init__
    self.step()
TypeError: step() missing 1 required positional argument: 'step'

Similar errors have been reported in the past, but the issue has not been resolved.  The line super().__init__(optimizer) in scheduler.py might make the cause of error.  In class CyclicLR(_LRScheduler), when "super().__init__(optimizer)" is executed, def __init__(self, optimizer, last_epoch=-1) of class _LRScheduler(object) in lr_scheduler.py executes self.step(). The above out that  self.step() in  class _LRScheduler(object) and def step(self, step) in class CyclicLR(_LRScheduler)are mistaken and confusing. Would you please helping me fixing abpbe error?

class CyclicLR(_LRScheduler):
def __init__(
            self, optimizer, base_lr=0.001, max_lr=0.006, step_size=2000.,
            mode='triangular', gamma=1., scale_fn=None, scale_mode='cycle'
    ):
        self.base_lr = base_lr
        self.max_lr = max_lr
        self.step_size = step_size
        self.mode = mode
        self.gamma = gamma
        if scale_fn == None:
            if self.mode == 'triangular':
                self.scale_fn = lambda x: 1.
                self.scale_mode = 'cycle'
            elif self.mode == 'triangular2':
                self.scale_fn = lambda x: 1/(2.**(x-1))
                self.scale_mode = 'cycle'
            elif self.mode == 'exp_range':
                self.scale_fn = lambda x: gamma**(x)
                self.scale_mode = 'iterations'
        else:
            self.scale_fn = scale_fn
            self.scale_mode = scale_mode
        self.clr_iterations = 0.
        self.trn_iterations = 0.
        self.history = {}
        self._reset()
        super().__init__(optimizer)

**Comment by Masanori-Yoshihira** [comment]
in class CyclicLR(_LRScheduler), when you made the method of "def step(self, step)", you forgot to set up "=None". When I fixed as "def step(self, step=None)", "$ python3 -m train.train_ycb_pvn3d" started to run without stop. But my workstation gpu memory was consumed all by your code and then stopped. I might have to fix more.

**Comment by ethnhe** [comment]
There are some changes on the ```_LRScheduler``` class after Pytorch version 1.0. If you are using a version higher than that, use the CyclicLR in the ```torch.optim.lr_scheduler``` module ([link1](https://github.com/ethnhe/PVN3D/blob/1fa7f0e9a03daa1b3f6ecf3d8ee819df322ff957/pvn3d/train/train_linemod_pvn3d.py#L27), [link2](https://github.com/ethnhe/PVN3D/blob/1fa7f0e9a03daa1b3f6ecf3d8ee819df322ff957/pvn3d/train/train_linemod_pvn3d.py#L486)) as in our [pytorch-1.5](https://github.com/ethnhe/PVN3D/tree/pytorch-1.5) branch.

**Comment by Masanori-Yoshihira** [comment]
to change torch.optim.lr_scheduler is not good.

**Comment by Masanori-Yoshihira** [comment]
in this case, how to make CyclicLR and how to override step function of torch.optim.lr_scheduler have problem. in CyclicLR, can i change from step(self, step) to step(self. step=None) or step(self. step=0)?

in case of step(self. step=None), clr(self, step) has an error which is cycle = np.floor(1+self.clr_iterations/(2*self.step_size)).

in case of  step(self. step=0), complex statement is shown on display like "RuntimeError: CUDA out of memory. Tried to allocate 226.00 MiB (GPU 0; 7.93 GiB total capacity; 6.93 GiB already allocated; 174.19 MiB free; 7.04 GiB reserved in total by PyTorch)".

I followed next step you wrote. pytorch version is not problem. there is only bug in your code.


    Install the NVIDIA driver version 440

    sudo apt install nvidia-driver-440
    # reboot
    shutdown -r now

    Install CUDA10.2 from the NIVIDA website. Remove other versions as explained in the NVIDIA installation guide. For Ubuntu 18.04 the instruction for "deb (network)" work well:

    wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64/cuda-ubuntu1804.pin
    sudo mv cuda-ubuntu1804.pin /etc/apt/preferences.d/cuda-repository-pin-600
    sudo apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64/7fa2af80.pub
    sudo add-apt-repository "deb http://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64/ /"
    sudo apt-get update
    sudo apt-get -y install cuda-10-2

    Create and activate a python3 virtual environment. Install the requirements.

    python3 -m venv venv
    source venv/bin/activate
    # Cython an numpy must be installed before requirements.txt
    pip3 install Cython numpy --user
    pip3 install -r requirements.txt

    Install tkinter through sudo apt install python3-tk
    Install python-pcl. For Ubtuntu 18.04 from source:

    # install dependencies
    sudo apt install libpcl-dev libvtk6-dev
    # clone fork with fix for Ubuntu 18.04
    git clone https://github.com/Tuebel/python-pcl
    cd python-pcl
    python3 setup.py install

    Install PointNet++ (refer from Pointnet2_PyTorch):

    git clone https://github.com/erikwijmans/Pointnet2_PyTorch
    cd Pointnet2_PyTorch
    pip3 install -r requirements.txt
    cd ..
    python3 setup.py build_ext

**Comment by Masanori-Yoshihira** [comment]
In torch1.0, torch.optim.lr_scheduler initialized the step variable of the step function in the initilizer; in torch1.5, it does not initialize the step variable of the step function in the initilizer. Your code should correctly initialize the step variable in the step function of the CyclicLR class, but that is not fixed. It also contains another new bug that uses up memory on the graphics board. No one will ever be able to run your pytorch1.5 branch as it is now.

**Comment by ethnhe** [comment]
I tested the torch-1.5 branch and it's working well. If you use pytorch1.5, just use the CyclicLR implemented by PyTorch [link](https://pytorch.org/docs/stable/optim.html?highlight=cycliclr#torch.optim.lr_scheduler.CyclicLR), as we did in the pytorch-1.5 branch, not the one we implemented in ```lib/utils/clr_callback.py```, which works well in torch=1.0 while not in version higher than 1.0.

**Comment by Masanori-Yoshihira** [comment]
i will try torch=1.0. but your torch-1.5 have reason why it never work. Regardless of the installation method, the relationship between CyclicLR and torch.optim.lr_scheduler in 1.5  is the same. and more, your pointnet2_utils folder does not contain _ext. your code must be changed from "from lib.pointnet2_utils import _ext"
to "import pointnet2_ops._ext as _ext",  futhermore, __init__.py should remove the line of "from . import _ext". Maybe, I recommend you find out how to install the libraries you need, starting from the fact that none of them are installed.

**Comment by ethnhe** [comment]
Thanks for the advice, I'll check them when I have more spare time.

**Comment by Masanori-Yoshihira** [comment]
The _ext and step() issue have been asked several times in the past by difference persons, but you have not provided any specific answers and realistic fixes. It seems to be as if you avoid to give direct answers.

**Comment by Masanori-Yoshihira** [comment]
However, I would like to thank you for making such new research results openly available. I suppose you can't take the time to fix things.

**Comment by ethnhe** [comment]
> The _ext and step() issue have been asked several times in the past by difference persons, but you have not provided any specific answers and realistic fixes. It seems to be as if you avoid to give direct answers.

In fact, the problem is only caused by torch version > 1.0 and the problem in the ```train_linemod_pvn3d.py``` have been fixed in the PyTorch-1.5 branch several months ago. As the first two links (link1, link2) I've provided before:

> There are some changes on the `_LRScheduler` class after Pytorch version 1.0. If you are using a version higher than that, use the CyclicLR in the `torch.optim.lr_scheduler` module ([link1](https://github.com/ethnhe/PVN3D/blob/1fa7f0e9a03daa1b3f6ecf3d8ee819df322ff957/pvn3d/train/train_linemod_pvn3d.py#L27), [link2](https://github.com/ethnhe/PVN3D/blob/1fa7f0e9a03daa1b3f6ecf3d8ee819df322ff957/pvn3d/train/train_linemod_pvn3d.py#L486)) as in our [pytorch-1.5](https://github.com/ethnhe/PVN3D/tree/pytorch-1.5) branch.

While the script ```train_ycb_pvn3d.py``` wasn't updated in that branch and I'll update it.

**Comment by ethnhe** [comment]
> your pointnet2_utils folder does not contain _ext. your code must be changed from "from lib.pointnet2_utils import _ext"
> to "import pointnet2_ops._ext as _ext", futhermore, **init**.py should remove the line of "from . import _ext". Maybe, I recommend you find out how to install the libraries you need, starting from the fact that none of them are installed.

If you follow the instruction in our readme and build the pointnet2 ext using scripts in ```PVN3D/setup.py``` rather than those in ```Pointnet2_PyTorch```, one copy of the built _ext*.so will be copied to ```pointnet2_utils folder```, as show in the [code](https://github.com/ethnhe/PVN3D/blob/1fa7f0e9a03daa1b3f6ecf3d8ee819df322ff957/setup.py#L35)
```
cd Pointnet2_PyTorch
pip3 install -r requirements.txt
cd ..
python3 setup.py build_ext
```
